xonsh/tools.py

```
@@ -956,19 +923,27 @@ def bool_or_int_to_str(x):
     return bool_to_str(x) if is_bool(x) else str(x)
 
 
-def ensure_int_or_slice(x):
-    """Makes sure that x is list-indexable."""
+
+@lazyobject
+def SLICE_REG():
+    return re.compile(r'(?P<start>(?:-\d)?\d*):(?P<end>(?:-\d)?\d*):?(?P<step>(?:-\d)?\d*)')
+
+def ensure_slice(x):
+    s = None
     if x is None:
         return slice(None)
-    elif is_int(x) or is_slice(x):
-        return x
-    # must have a string from here on
-    if is_slice_as_str(x):
+    try:
+        x = int(x)
+        s = slice(x, x+1)
+    except ValueError:
         x = x.strip('[]()')
-        return slice(*(int(x) if len(x) > 0 else None for x in x.split(':')))
-    elif is_int_as_str(x):
-        return int(x)
-    return False
+        m = SLICE_REG.fullmatch(x)
+        if m:
+            groups = (int(i) if i else None for i in m.groups())
+            s = slice(*groups)
+    except TypeError:
+        pass
+    return s
```

Ensurers should raise an error if they fail.